### PR TITLE
[Fix] Support single cahnnel `pred` for Binary Cross Entropy Loss

### PR DESCRIPTION
**Pull request by RockeyCoss**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix the bug that binary cross entropy loss doesn't support single channel input.
Related Issue: https://github.com/open-mmlab/mmsegmentation/issues/1430

## Modification

1. When the channel of `pred` is one, instead of expanding the lebel to one-hot format, `binary_cross_entropy_loss` only unsqueeze the `label` to match the shape of the `pred`.
2. Add a new test function to cover the new codes.

### CODE REVIEW

File: mmseg/models/losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -120,10 +120,21 @@ def binary_cross_entropy(pred,
                 pred.dim() == 4 and label.dim() == 3), \
             'Only pred shape [N, C], label shape [N] or pred shape [N, C, ' \
             'H, W], label shape [N, H, W] are supported'
-        # `weight` returned from `_expand_onehot_labels`
-        # has been treated for valid (non-ignore) pixels
-        label, weight, valid_mask = _expand_onehot_labels(
-            label, weight, pred.shape, ignore_index)
+
+        if pred.size(1) == 1:
```
Comment (RockeyCoss) [inline mmseg/models/losses/cross_entropy_loss.py:124]: do we need to judge whether `torch.max(label) == 1` here?

**Review by MeowZheng — COMMENTED** [review]
Please refine pr tittle and pr msg first ref #1438

File: mmseg/models/losses/cross_entropy_loss.py [hunk 2]
```diff
@@ -120,10 +120,21 @@ def binary_cross_entropy(pred,
...
+            # For binary class segmentation, the shape of `pred` is
+            # [N, 1, H, W] and that of `label` is [N, H, W].
+            label = label.unsqueeze(1)
+            valid_mask = ((label >= 0) & (label != ignore_index)).float()
+            if weight is not None:
+                weight = weight.unsqueeze(1).float() * valid_mask
+            else:
+                weight = valid_mask
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:132]: I think moving these lines to `_expand_onehot_labels` might be better, as the function of these lines is expanding labels.

File: mmseg/models/losses/cross_entropy_loss.py [hunk 3]
```diff
@@ -65,23 +65,33 @@ def cross_entropy(pred,
...
     """Expand onehot labels to match the size of prediction."""
-    bin_labels = labels.new_zeros(target_shape)
-    valid_mask = (labels >= 0) & (labels != ignore_index)
-    inds = torch.nonzero(valid_mask, as_tuple=True)
-
-    if inds[0].numel() > 0:
-        if labels.dim() == 3:
-            bin_labels[inds[0], labels[valid_mask], inds[1], inds[2]] = 1
+    if target_shape[1] == 1:
+        # For binary class segmentation, the shape of `pred` is
```
Comment (MeowZheng) [inline mmseg/models/losses/cross_entropy_loss.py:69]: I think it needs to add assertion like
```python
assert label.max() <= 1, 'balabala'
```

**Review by MengzhangLI — APPROVED** [review]
